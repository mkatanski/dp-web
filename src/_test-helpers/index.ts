export * from "./withResponsiveContext";
export * from "./withThemeProvider";
