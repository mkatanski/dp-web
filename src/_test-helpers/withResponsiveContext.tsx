import React from "react";
import { Context as ResponsiveContext } from "react-responsive";
import { Breakpoints } from "config/breakpoints";

export const withMobileSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.mobile }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withTabletSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.tablet }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withDesktopSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.desktop }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withLargeDesktopSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.largeDesktop }}>
    {Component}
  </ResponsiveContext.Provider>
);
