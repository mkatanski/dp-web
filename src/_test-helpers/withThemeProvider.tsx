import React from "react";
import { ThemeProvider as DefaultThemeProvider } from "lib/theme/ThemeProvider";
import { StyledTheme, Theme } from "styles/theme/_types";
import { createMuiTheme } from "@material-ui/core";

const testTheme: StyledTheme = {
  primary: "primary",
  secondary: "secondary",
  textPrimary: "textPrimary",
  textSecondary: "textSecondary",
  textInverse: "textInverse"
};

const muiTheme = createMuiTheme({
  direction: "ltr"
});

const theme: Theme = Object.assign(testTheme, muiTheme);

export const withThemeProvider = (Component: React.ReactNode) => (
  <DefaultThemeProvider theme={theme}>{Component}</DefaultThemeProvider>
);
