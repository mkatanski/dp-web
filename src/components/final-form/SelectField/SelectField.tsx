import React from "react";
import { SelectFieldView, SelectFieldViewProps } from "./SelectField.view";
import { useField } from "react-final-form";

// We use omit to filter out properties of components view
// which we don't want to expose
// See: https://www.typescriptlang.org/docs/handbook/utility-types.html#omittk
// for usage example
export type SelectFieldProps = Omit<
  SelectFieldViewProps,
  "value" | "onChange" | "value" | "errorMessage" | "error"
> & {};

export const SelectField: React.FC<SelectFieldProps> = ({
  children,
  name,
  ...rest
}: SelectFieldProps) => {
  const { input, meta } = useField(name, { parse: v => v });

  const handleOnChange = (value: unknown) => {
    input.onChange(value);
  };

  return (
    <SelectFieldView
      {...rest}
      name={name}
      value={input.value}
      onChange={handleOnChange}
      errorMessage={meta.error}
    >
      {children}
    </SelectFieldView>
  );
};
