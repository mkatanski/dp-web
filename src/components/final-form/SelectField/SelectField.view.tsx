import React from "react";
import {
  FormControl as MuiFormControl,
  InputLabel,
  Select,
  FormHelperText,
  SelectProps
} from "@material-ui/core";
import styled from "styled-components";

export type SelectFieldViewProps = SelectProps & {
  name: string;
  value: unknown;
  label?: React.ReactNode;
  id?: string;
  testId?: string;
  errorMessage?: string;
  onChange?: (value: unknown) => void;
};

const FormControl = styled(MuiFormControl)`
  min-width: 100px;
  width: 100%;
  margin: ${({ theme }) => theme.spacing(1, 0)};
`;

export const SelectFieldView: React.FC<SelectFieldViewProps> = ({
  children,
  label,
  name,
  testId = `rff-select-${name}`,
  id = name,
  value,
  errorMessage: error,
  onChange,
  variant,
  ...rest
}: SelectFieldViewProps) => {
  const handleChange: SelectProps["onChange"] = event => {
    onChange && onChange(event.target.value as string[]);
  };

  return (
    <FormControl variant={variant} error={!!error}>
      {label && <InputLabel id={`${id}-label`}>{label}</InputLabel>}
      <Select
        {...rest}
        labelId={`${id}-label`}
        id={id}
        name={name}
        data-testid={testId}
        value={value}
        onChange={handleChange}
        label={label}
        inputProps={{
          name,
          id,
          "data-testid": `${testId}-input`
        }}
      >
        {children}
      </Select>
      {error && <FormHelperText>{error}</FormHelperText>}
    </FormControl>
  );
};
