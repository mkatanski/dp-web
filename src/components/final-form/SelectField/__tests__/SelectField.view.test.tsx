import React from "react";
import TestRenderer from "react-test-renderer";
import "jest-styled-components";

import { SelectFieldView } from "../SelectField.view";
import { withThemeProvider } from "_test-helpers";

describe("components", () => {
  describe("SelectField", () => {
    describe("SelectFieldView", () => {
      it("should render SelectFieldView component", () => {
        const tree = TestRenderer.create(
          withThemeProvider(<SelectFieldView name="test" value="test" />)
        );
        expect(tree).toMatchSnapshot();
      });

      it("should render SelectFieldView component with children", () => {
        const tree = TestRenderer.create(
          withThemeProvider(
            <SelectFieldView name="test" value="test">
              child
            </SelectFieldView>
          )
        );
        expect(tree).toMatchSnapshot();
      });

      it("should render SelectFieldView with error", () => {
        const tree = TestRenderer.create(
          withThemeProvider(
            <SelectFieldView
              name="test2"
              value="test2"
              errorMessage="Something went wrong"
            />
          )
        );
        expect(tree).toMatchSnapshot();
      });

      it("should render SelectFieldView with label", () => {
        const tree = TestRenderer.create(
          withThemeProvider(
            <SelectFieldView name="test2" value="test2" label="This is label" />
          )
        );
        expect(tree).toMatchSnapshot();
      });
    });
  });
});
