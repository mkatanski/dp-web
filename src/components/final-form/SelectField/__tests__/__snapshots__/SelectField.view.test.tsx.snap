// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components SelectField SelectFieldView should render SelectFieldView component 1`] = `
.c0 {
  min-width: 100px;
  width: 100%;
  margin: 8px 0px;
}

<div
  className="c0"
  data-props={
    Object {
      "error": false,
      "variant": undefined,
    }
  }
  id="MUI.FormControl"
>
  <select
    data-props={
      Object {
        "data-testid": "rff-select-test",
        "id": "test",
        "inputProps": Object {
          "data-testid": "rff-select-test-input",
          "id": "test",
          "name": "test",
        },
        "label": undefined,
        "labelId": "test-label",
        "onChange": [Function],
        "value": "test",
      }
    }
    id="MUI.Select"
    name="test"
  />
</div>
`;

exports[`components SelectField SelectFieldView should render SelectFieldView component with children 1`] = `
.c0 {
  min-width: 100px;
  width: 100%;
  margin: 8px 0px;
}

<div
  className="c0"
  data-props={
    Object {
      "error": false,
      "variant": undefined,
    }
  }
  id="MUI.FormControl"
>
  <select
    data-props={
      Object {
        "data-testid": "rff-select-test",
        "id": "test",
        "inputProps": Object {
          "data-testid": "rff-select-test-input",
          "id": "test",
          "name": "test",
        },
        "label": undefined,
        "labelId": "test-label",
        "onChange": [Function],
        "value": "test",
      }
    }
    id="MUI.Select"
    name="test"
  >
    child
  </select>
</div>
`;

exports[`components SelectField SelectFieldView should render SelectFieldView with error 1`] = `
.c0 {
  min-width: 100px;
  width: 100%;
  margin: 8px 0px;
}

<div
  className="c0"
  data-props={
    Object {
      "error": undefined,
      "variant": undefined,
    }
  }
  id="MUI.FormControl"
>
  <select
    data-props={
      Object {
        "data-testid": "rff-select-test2",
        "id": "test2",
        "inputProps": Object {
          "data-testid": "rff-select-test2-input",
          "id": "test2",
          "name": "test2",
        },
        "label": undefined,
        "labelId": "test2-label",
        "onChange": [Function],
        "value": "test2",
      }
    }
    id="MUI.Select"
    name="test2"
  />
</div>
`;

exports[`components SelectField SelectFieldView should render SelectFieldView with label 1`] = `
.c0 {
  min-width: 100px;
  width: 100%;
  margin: 8px 0px;
}

<div
  className="c0"
  data-props={
    Object {
      "error": false,
      "variant": undefined,
    }
  }
  id="MUI.FormControl"
>
  <div
    data-props={
      Object {
        "id": "test2-label",
      }
    }
    id="MUI.InputLabel"
  >
    This is label
  </div>
  <select
    data-props={
      Object {
        "data-testid": "rff-select-test2",
        "id": "test2",
        "inputProps": Object {
          "data-testid": "rff-select-test2-input",
          "id": "test2",
          "name": "test2",
        },
        "label": "This is label",
        "labelId": "test2-label",
        "onChange": [Function],
        "value": "test2",
      }
    }
    id="MUI.Select"
    name="test2"
  />
</div>
`;
